lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+          Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+        }
+      }
+    }
+  }
+
+  // Elect a final stack offset.
+  const auto WinFSO = electFSO(MaybeReturns);
+
+  // Did we find a valid return instruction?
+  for (const auto &E : MaybeReturns) {
+    auto *CI = E.first;
+    if (E.second == *WinFSO) {
+      Result.insert({ E.first, FunctionEdgeTypeValue::Return });
+      unsigned NumArgs = CI->getNumArgOperands();
+      if (NumArgs > 3) {
+        for (unsigned Idx = 3; Idx < NumArgs; ++Idx) {
+          if (isa<ConstantInt>(CI->getArgOperand(Idx))) {
+            if (cast<ConstantInt>(CI->getArgOperand(Idx))->getZExtValue() == 0)
+              CalleeSavedRegs.insert(ABIRegisters[Idx - 3]);
+          }
+        }
+      }
+    } else {
+      Result.insert({ E.first, FunctionEdgeTypeValue::BrokenReturn });
+    }
+  }
+
+  // Neither a return nor a broken return?
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0)))
+        continue;
+
+      auto *V = dyn_cast<ConstantInt>(CI->getArgOperand(1));
+      if (WinFSO.has_value() && V != nullptr && V->getZExtValue() == *WinFSO)
+        Result.insert({ CI, FunctionEdgeTypeValue::IndirectTailCall });
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
+  } else {
+#ifdef NORETURN_IS_FIXED
+    Type = FunctionTypeValue::NoReturn;
+#else
+    Type = FunctionTypeValue::Regular;
+#endif
+  }
+
+  // Finally retrieve the clobbered registers and return them.
+  std::erase_if(ClobberedRegs, [&](const auto &E) {
+    return CalleeSavedRegs.find(E) != CalleeSavedRegs.end();
+  });
+
+  // Finalize CFG for the model.
+  for (const auto &[CI, EdgeValue] : Result) {
+    auto PC = MetaAddress::fromConstant(CI->getArgOperand(2));
+    auto &Block = CFG.at(PC);
+    Block.Successors.insert(makeEdge(MetaAddress::invalid(), EdgeValue));
+  }
+
+  // We say that a register is callee-saved when, besides being preserved by the
+  // callee, there is at least a write onto this register.
+  std::set<GlobalVariable *> ActualCalleeSavedRegs;
+  std::set_intersection(CalleeSavedRegs.begin(),
+                        CalleeSavedRegs.end(),
+                        WrittenRegisters.begin(),
+                        WrittenRegisters.end(),
+                        std::inserter(ActualCalleeSavedRegs,
+                                      ActualCalleeSavedRegs.begin()));
```

Comment:
Maybe make an helper function:
```suggestion
  std::set<GlobalVariable *> ActualCalleeSavedRegs = intersect(CalleeSavedRegs, WrittenRegisters);
```

Reply:
Why is this marked as solved? It's not!